### DISCUSSION TITLE
Getting 403 error even with User-Agent header

### DISCUSSION THREAD

**Discussion by MalenaID** [weather-gov/api#673] [Q&A]
A training used the API to get station observations. About a month ago, we started getting a 403 errors. Nothing on our end on how to access the API changed. 

We used Postman to test out the API request and it works there.  This is the error message:

`  [
  {
    "code": 403,
    "message": "<HTML><HEAD>\n<TITLE>Access Denied</TITLE>\n</HEAD><BODY>\n<H1>Access Denied</H1>\n \nYou don't have permission to access \"http&#58;&#47;&#47;api&#46;weather&#46;gov&#47;stations&#47;KSFO&#47;observations&#47;latest\" on this server.<P>\nReference&#32;&#35;18&#46;dc83017&#46;1700708901&#46;417077d9\n</BODY>\n</HTML>\n",
    "occurredAt": 1700708901636,
    "resolved": false,
    "source": "application",
    "classification": "connection"
  }
]`

Any ideas on what could cause the error? I tried different User-Agents to see if that might make a difference. It didn't. 

I don't know how many people keep the training exercise active, so I don't know if it's realistic to suspect maybe there were too many API requests, which may have caused this. Any insight is appreciated.

**Comment by dwhitemv25** [comment]
Use `https://` instead of `http://`.

**Reply by MalenaID** [reply]
Huh. Thanks for pointing that out. I am using https:// when trying. I wonder why the message states I'm not.

**Reply by sullynole** [reply]
Are you still seeing this?  Might be a temporary block due to excessive requests.

**Reply by MalenaID** [reply]
We've been seeing this for at least a month.

**Reply by dwhitemv25** [reply]
I did a test and api.weather.gov issues a 301 redirect to HTTPS if you access it with HTTP, so it's not that. 

Might need to send the reference # to NCO Ops and have them look up the reason for the error.

Can you reproduce the problem with any other tool at your site (i.e., web browser, `curl`)?

Does the Reference # change each time you access it? If not, you might have a proxy interfering with your request and serving a (stale) error message.  

Is there any way you can perform a request and capture the headers sent and received?

**Reply by MalenaID** [reply]
Hello. Yes, the reference numbers change each time. I can see the headers sent. It doesn't show anything, even though I have a user-agent header. It doesn't appear in the header section of the request. And no headers appear on the response.  I tried a browser and I get the correct response there.

**Comment by scadergit** [comment]
Sorry for the late reply, I can look up the code, but I need a new one as the original one you provided has expired.

**Reply by MalenaID** [reply]
The reference codes? Thanks. If so, I have two new ones: Reference&#32;&#35;18&#46;89c83017&#46;1701292793&#46;2ec536e4

Reference&#32;&#35;18&#46;89c83017&#46;1701292863&#46;2ed086d8

`[
  {
    "code": 403,
    "message": "<HTML><HEAD>\n<TITLE>Access Denied</TITLE>\n</HEAD><BODY>\n<H1>Access Denied</H1>\n \nYou don't have permission to access \"http&#58;&#47;&#47;api&#46;weather&#46;gov&#47;zones&#47;forecast&#47;TXZ211&#47;observations\" on this server.<P>\nReference&#32;&#35;18&#46;89c83017&#46;1701292863&#46;2ed086d8\n</BODY>\n</HTML>\n",
    "occurredAt": 1701292863913,
    "resolved": false,
    "source": "application",
    "classification": "connection"
  }
]`

**Reply by scadergit** [reply]
Hmm, the first time I've seen this, but appears your requests are getting tagged as bot requests--often times due to developer frameworks, which ahere to strict rate policies. Odd given you've tried other user agents. I would reach out to NCO to investigate further.

**Reply by MalenaID** [reply]
Thank you very much for checking. I emailed them.

**Comment by joepperkins** [comment]
You gateway IP address has been blocked at the WAF, try using a VPN or proxy to see if the issue persists.  To many requests or bad quests can cause this blocking and can be a result of the backend service, for example you send a valid request but the backend service respondes with a non 200 response repeatedly can trigger this.

**Comment by MalenaID** [comment]
Thanks. I've asked support twice if we were being blocked and I kept
getting responses that it's because of the (not) blank user-agent. This is
an integration software being used to connect. I can't tell all the users
to use a proxy or VPN.

UPDATE: And now that someone else checked, the request are working fine. I'll regret still not knowing the cause and solution, but I'll be happy that it's working for now.

Thanks everyone for your help!
